omegaconf/basecontainer.py

```diff
@@ -229,9 +229,13 @@ def convert(val: Node) -> Any:
                 node = conf._get_node(key, throw_on_missing_value=throw_on_missing)
                 assert isinstance(node, Node)
                 if resolve:
-                    node = node._dereference_node(
-                        throw_on_missing=False, throw_on_resolution_failure=True
-                    )
+                    try:
+                        node = node._dereference_node(
+                            throw_on_missing=False, throw_on_resolution_failure=True
+                        )
+                    except MissingMandatoryValue as e:
+                        assert node is not None
+                        conf._format_and_raise(key=key, value=node._value(), cause=e)
```

Comment:
1. Why is _dereference_node throwing a MandatoryMissingValue is throw_on_missing is False?
2. Why are you passing False to throw_on_missing and not the input flag?

Reply:
> 1. Why is _dereference_node throwing a MandatoryMissingValue is throw_on_missing is False?

This is because we are calling `_dereference_node` with `throw_on_resolution_failure=True`.
If #545 is merged in it's current state, `_dereference_node` will raise `MandatoryMissingValue` if `throw_on_resolution_failure==True` and the interpolation cannot be resolved.

This being said, I actually think it makes more sense to change this from
```python
_dereference_node(..., throw_on_resolution_failure=True)
```
to
```python
_dereference_node(..., throw_on_resolution_failure=throw_on_missing)
```
so that resolution failure will not cause error if the user does not specify the `throw_on_missing` option. I'll add tests to this effect...

Reply:
> 2. Why are you passing False to throw_on_missing and not the input flag?

The current workflow is:
- first, get the node and (optionally) throw if missing:
```python
node = conf._get_node(key, throw_on_missing_value=throw_on_missing)
```
- second: if `resolve==True`, dereference the node:
```python
node = node._dereference_node(throw_on_missing=False, throw_on_resolution_failure=True)
```
Since the `_get_node` call already has `throw_on_missing_value=throw_on_missing`, it would be redundant to use the `throw_on_missing` flag in the `_dereference_node` call as well.